### ISSUE TITLE
Code Size: Driver Wrappers: Prototype mechanism

### ISSUE THREAD

**Issue by tom-cosgrove-arm** [Mbed-TLS/mbedtls#7395]
When `psa_crypto_driver_wrappers.c` is created from `psa_crypto_driver_wrappers.c.jinja` but neither `PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT` nor `PSA_CRYPTO_DRIVER_TEST` are set, it should be possible to omit the entire wrapper function.

This probably requires the header for the driver wrappers to define the driver wrapper as the fallback function, and for the C code to omit this function, under those conditions, but other implementation possibilities might be linker aliases or a `static inline` function wrapper.

This ticket is to prototype the mechanism for a small number (2-3) of functions.

**Comment by gilles-peskine-arm** [comment]
My hunch is that the compiler will do it for us if we `#include "psa_crypto_driver_wrappers.c"` and make the functions static. (Link-time optimization should also do the trick, but not everyone can build with LTO.) We'd gain more by putting the built-in implementations in the same compilation unit, but that would be structurally hard to manage, and there I don't think the gains are worth the trouble.

**Comment by gilles-peskine-arm** [comment]
> My hunch is that the compiler will do it for us if we #include "psa_crypto_driver_wrappers.c" and make the functions static

Not quite: that won't do everything.

There's at least one thing the compiler won't be able to optimize that way, which is the location check. During key creation, we need `if (location != 0) return PSA_ERROR_INVALID_ARGUMENT`, and that's what the generic code does if there are no drivers. But when accessing an existing key, if there are no drivers (in fact, if even there are accelerator drivers, as long as there are no SE drivers), then the generic code will still do a location check which boils down to `if (location != 0) return PSA_ERROR_INVALID_ARGUMENT`. but this can be completely removed since we know that keys are only created at location 0.

However, we might want to keep that check anyway to be more robust against a misconfiguration, if an installation with drivers is updated to remove all SE drivers and there are keys in storage with `location!=0`.

**Comment by xkqian** [comment]
I have some tests locally with the tool `mbedtls-docs/tools/bin/mtest` and get the following result of `libmbedcrypto.a`: I just treat them into 3 kinds: 1: `Original code` from latest branch. 2: Define the functions in the file scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja into "`static inline`" except the function psa_driver_wrapper_export_public_key(). 3. Add macro guards to make the function call directly into the built-in functions, like :"```
#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT) || \
    defined(PSA_CRYPTO_DRIVER_TEST)
psa_driver_wrapper_sign_message()
{
    ...
}
#else   
#define psa_driver_wrapper_sign_message psa_sign_message_builtin
#endif 
```"
The result is listed as below:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/xiaqia01/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/xiaqia01/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT | PSA_CRYPTO_DRIVER_TEST | Original(1) | Static inline(2) | Static inline with macro guard(3)
-- | -- | -- | -- | --
Disabled | Disabled | 74174 | 72414 | 72414
Enabled | Disabled | 70248 | 68372 | 68372



</body>

</html>

**Comment by xkqian** [comment]
We can see,compared with the original code, if we define the functions to "`static inline`", the code size will be decreased when we disabled the macro  `PSA_CRYPTO_DRIVER_TEST`.  It's decreased by `1760 bytes` when PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT is diabled and `1876 bytes` when it's enabled.   And in the two cases, if we have defined them into `static inline`, then the` directly define into the built-in functions `isn't helpful further to the code size.
